### PR TITLE
feat: Add implementation and testing

### PR DESCRIPTION
**Pull request by lucas-manuel**

### CODE REVIEW

File: src/ERC20.sol [hunk 1]
```diff
@@ -0,0 +1,83 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.6;
```
Comment (JGcarv) [inline src/ERC20.sol]: Shouldn't we choose a fixed compiler?

File: src/ERC20.sol [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    mapping(address => mapping(address => uint256)) public override allowance;
+
+    constructor(string memory _name, string memory _symbol, uint8 _decimals) {
+        name     = _name;
+        symbol   = _symbol;
+        decimals = _decimals;
+    }
+
+    /*********************************/
+    /*** ERC-20 External Functions ***/
```
Comment (JGcarv) [inline src/ERC20.sol]: Does the `ERC-20` prefix is needed here?

File: src/ERC20.sol [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+    function increaseAllowance(address spender, uint256 amount) external override returns (bool) {
+        _approve(spender, allowance[msg.sender][spender] + amount);
+        return true;
+    }
+
+    function decreaseAllowance(address spender, uint256 amount) external override returns (bool) {
+        _approve(spender, allowance[msg.sender][spender] - amount);
+        return true;
+    }
```
Comment (JGcarv) [inline src/ERC20.sol:42]: Shouldn't this be inverted to follow the alphabetical order?

File: src/IERC20.sol [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    /**
+     * @dev   Moves a pre-approved amount of tokens from a sender to a specified account.
+     * @dev   Emits a {Transfer} event.
+     * @param owner     Account that tokens are moving from.
+     * @param recipient Account that recieves tokens.
+     * @param amount    Amount of tokens that are transferred.
+     */
+    function transferFrom(address owner, address recipient, uint256 amount) external returns (bool);
+
+}
```
Comment (JGcarv) [inline src/IERC20.sol]: These are all following semantic order, not alphabetical. I honestly think this way it makes more sense, but our style guide asks for alphabetical, right?

**Review by JGcarv — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 5]
```diff
@@ -0,0 +1 @@
+/out
```
Comment (deluca-mike) [inline .gitignore:1]: Use our "common" overhead.

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,4 @@
+all    :; dapp build
+clean  :; dapp clean
+test   :; dapp test --verbosity 2
+deploy :; dapp create Erc20
```
Comment (deluca-mike) [inline Makefile]: Use our "common" overhead.

File: src/ERC20.sol (hunk 2 above)
Comment (deluca-mike) [reply src/ERC20.sol]: Agreed. Remove.

File: src/ERC20.sol [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+    function transferFrom(address owner, address recipient, uint256 amount) external override returns (bool) {
+        allowance[owner][msg.sender] -= amount;
+        _transfer(owner, recipient, amount);
+        return true;
+    }
+
+    /*********************************/
+    /*** ERC-20 Internal Functions ***/
```
Comment (deluca-mike) [inline src/ERC20.sol]:
```suggestion
    /*** Internal Functions ***/
```

**Review by deluca-mike — CHANGES_REQUESTED** [review]

**Review by deluca-mike — APPROVED** [review]

**Review by deluca-mike — CHANGES_REQUESTED** [review]
ERC20User is not functional

File: src/test/accounts/ERC20User.sol [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+import { ERC20 } from "../../ERC20.sol";
+
+contract ERC20User {
+
+    ERC20 token;
+
+    constructor(ERC20 _token) {
+        token = _token;
+    }
```
Comment (deluca-mike) [inline src/test/accounts/ERC20User.sol]: Scrap this in favour of functional design.

File: src/test/accounts/ERC20User.sol [hunk 9]
```diff
@@ -0,0 +1,44 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.6;
+
+import { ERC20 } from "../../ERC20.sol";
```
Comment (deluca-mike) [inline src/test/accounts/ERC20User.sol]:
```suggestion
import { IERC20 } from "../../interfaces/IERC20.sol";
```

File: src/test/accounts/ERC20User.sol [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    constructor(ERC20 _token) {
+        token = _token;
+    }
+
+    /************************/
+    /*** Direct Functions ***/
+    /************************/
+    function erc20_approve(address recipient, uint256 amount) external {
+        token.approve(recipient, amount);
+    }
```
Comment (deluca-mike) [inline src/test/accounts/ERC20User.sol]:
```suggestion
    function erc20_approve(address token, address recipient, uint256 amount) external {
        IERC20(token).approve(recipient, amount);
    }
```

File: src/test/accounts/ERC20User.sol [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+    function erc20_transfer(address recipient, uint256 amount) external {
+        token.transfer(recipient, amount);
+    }
```
Comment (deluca-mike) [inline src/test/accounts/ERC20User.sol]: Ditto

File: src/test/accounts/ERC20User.sol [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+    function erc20_transferFrom(address owner, address recipient, uint256 amount) external {
+        token.transferFrom(owner, recipient, amount);
+    }
```
Comment (deluca-mike) [inline src/test/accounts/ERC20User.sol]: Ditto.

File: src/test/accounts/ERC20User.sol [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+
+    /************************/
+    /*** Try Functions ***/
+    /************************/
+    function try_erc20_approve(address recipient, uint256 amount) external {
+        token.approve(recipient, amount);
+    }
```
Comment (deluca-mike) [inline src/test/accounts/ERC20User.sol]:
```suggestion
    function try_erc20_approve(address token, address recipient, uint256 amount) external returns (bool ok) {
        (ok,) = token.call(abi.encodeWithSelector(IERC20.approve.selector, recipient, amount));
    }
```

File: src/test/accounts/ERC20User.sol [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+    function try_erc20_transfer(address recipient, uint256 amount) external {
+        token.transfer(recipient, amount);
+    }
```
Comment (deluca-mike) [inline src/test/accounts/ERC20User.sol]: Ditto.

File: src/test/accounts/ERC20User.sol [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+
+    function try_erc20_transferFrom( address owner, address recipient, uint256 amount) external {
+        token.transferFrom(owner, recipient, amount);
+    }
```
Comment (deluca-mike) [inline src/test/accounts/ERC20User.sol]: Ditto.

**Review by deluca-mike — CHANGES_REQUESTED** [review]
Needs a more functional `ERC20User`

**Review by deluca-mike — APPROVED** [review]

**Pull request merged**